DEV-37: Lint custom javascript with eslint

jesconstantine opened the pull request:
This PR:

* adds a new phing target for installing npm dependencies for the project and for Drupal core
* adds this new phing target as a dependency for the existing `code-review` target
* adds eslint as a step to the `code-review` target
* removes `.js` files from `phpmd.xml` to ensure PHP linters are linting js files

Test this PR with https://github.com/palantirnet/drupal-skeleton/pull/146

jesconstantine committed:
Exclude .js from php linters

jesconstantine committed:
Install node dependencies and lint custom js in code-review target

jesconstantine commented on defaults/install/build.xml:
```diff
@@ -127,6 +127,10 @@
+
+        <!-- Run eslint on custom javascript -->
+        <echo msg="$> npx eslint --ext .js ${drupal.root}/modules/custom" />
+        <exec command="npx eslint --ext .js --no-error-on-unmatched-pattern ${drupal.root}/modules/custom" logoutput="true" checkreturn="true" />
```
Should we also be linting custom theme files?

agentrickard replied:
It depends -- those are generally in a styleguide which should handle its own linting. 

Perhaps a config option in `build.yml` to include theme linting?

agentrickard replied:
A config is what I did in DHS -- https://github.com/palantirnet/wisdhs-workweb/pull/411/files

jesconstantine committed:
Add config options for eslint directories and extensions

jesconstantine committed:
Use core's actual eslint dependency and our new eslint config options

jesconstantine committed:
Revert "Use core's actual eslint dependency and our new eslint config options"

This reverts commit 53cbedb813adda654d032dda16a482022e44cb84.

jesconstantine committed:
Use built in tools for running eslint (node) and checking dependencies (yarn)

jesconstantine committed:
TEMP: update acquia memcache build install steps based on their docs

https://docs.acquia.com/cloud-platform/performance/memcached/enable/

The old file is no longer available and causes the-build-installer to fail

jesconstantine committed:
Use yarn to execute eslint binary

jesconstantine committed:
remove dependency on node dependencies from code-review task

we add it to the new eslint task instead

jesconstantine committed:
create discrete eslint phing task

jesconstantine committed:
use recommended yarn cli command for checking installed packages

see note: https://classic.yarnpkg.com/en/docs/cli/check#toc-yarn-check

jesconstantine committed:
Don't touch settings file which is being removed anyway

This was a temporary and unnecessary change

jesconstantine commented on targets/install.xml:
```diff
@@ -280,7 +280,14 @@
                 <!-- Site-level setup -->
                 <!-- @see https://docs.acquia.com/cloud-platform/performance/memcached/enable/ -->
-                <httpget url="https://docs.acquia.com/_downloads/cloud-memcache-d8+.php" dir="${build.dir}/${drupal.root}/sites/${drupal.site.dir}" filename="settings.acquia-memcache.php"/>
+                <composer command="require" composer="${composer.composer}">
+                    <arg line="--working-dir ${application.startdir}"/>
+                    <arg value="drupal/memcache"/>
+                </composer>
+                <composer command="require" composer="${composer.composer}">
+                    <arg line="--working-dir ${application.startdir}"/>
+                    <arg value="acquia/memcache-settings"/>
+                </composer>
```
We can roll this back once https://github.com/palantirnet/the-build/pull/226 is merged but this unblocks the build installer test step.

jesconstantine committed:
Clean up comments for the eslint target

jesconstantine committed:
Make the directions clearer for multiple eslint directory patterns

jesconstantine committed:
Add more context for eslint ext argument.

jesconstantine committed:
I think we should allow ignore files, patterns to be used.

For example, custom modules might come with their own eslint setup.

patrickfweston reviewed (approved):
LGTM and the change in the `drupal-skeleton` project looks great!

agentrickard committed:
Merge branch 'develop' into dev-37-eslint

Commit:
dev-37 - Use php 8.1.18.

Commit:
Revert "dev-37 - Use php 8.1.18."

This reverts commit 893856df3621f4c645a6d6e3c6f1d5186ff4f7d4.

byrond committed:
Merge branch 'develop' into dev-37-eslint
